hedera-node/docs/design/services/smart-contract-service/hedera-account-service-isAuthorized-design.md

```diff
@@ -0,0 +1,88 @@
+`...contract.impl.exec.processors.HasTranslatorsModule` (for the individual methods).
+
+Following the existing pattern for HAS each method will be given a package under 
+`...contract.impl.exec.systemcontracts.has` with two classes:
+* `[method]Translator` - which first confirms that the call is for the given method, and
+  then builds a `TransactionBody` for it.
+* `[method]Call` - which takes that `TransactionBody` and then has an `execute()` method
+  that does the work.
+
+
+## Design for `isVirtualAddress`
+
+Looking up a Hedera account from an EVM address is done through the
+`HederaWorldUpdator.Enhancement` which can map from an EVM address to the Hedera account it
+is an alias for.  This functionality is provided by `HasCallAttempt.linkedAccount()`.
+
+## Design for `isAuthorizedRaw`
+
+[TBD shortly as I fix up some problems with understanding HAS]
+
+`isAuthorizedRaw` has two flows: Either the argument address is a Hedera account, or it is
+an EVM address.
+
+* If an EVM address then this is simply the `ECRECOVER` precompile.  We have direct
+access to the `ECRECOVER` precompile from BESU in the class
+`org.hyperledger.besu.evm.precompile.ECRECPrecompiledContract` which will compute
+the `ECRECOVER` operation given the message hash + signature (encoded as a calldata for
+the precompile) (and a `MessageFrame` which it ignores).  This returns either the
+recovered address, if successful, or "empty", if not.  The recovered address must then
+match the address given as an argument to `isAuthorizedRaw`.
+  
+* If an Hedera account then it gets the account, looks in it to find a key _which it must
+have_ and which must be a _single key_ (otherwise: failure).  Then given the key and the
+signature verifies that the signature matches the message hash and is attested by 
+the account. This is an operation provided by `HandleContext` (which can be accessed via
+the `HederaWorldUpdater.Enhancement`).
+
+## TBD
+
+
+* `isAuthorizedRaw` requires a message hash.  For EVM addresses, with an EC key, 
+the message hash is a
+KECCAK256 which is provided directly by the EVM and is relatively inexpensive.  But for
+Hedera accounts, with an ED key, the message hash is a SHA384 which can, at this time,
+only be computed by an expensive EVM script - on the order of ~1.5-2M gas.  We should
+provide a way for the contract to compute this hash at the same level of expense as the
+KECCAK256 version.
+  
+* `isAuthorized{Raw}` does a relatively expensive (resource-wise) computation that should
+be charged for.  Currently we have no method for charging fees (gas or otherwise) for
+system contract calls, not to mention varying the fee for individual system contract
+methods.  (All existing system contract methods are either "cheap" - as in too small to
+charge for - or involve child transactions that themselves do incur fees.) 
+  * TBD #1: A
+mechanism to charge fees for individual system contract methods 
+  * TBD #2: How do we
```

Comment:
matching ecrecover is a good starting point. It works for the EC case but an additional buffer may be needed for the node logic.
We should charge at least the amount of an HCS transaction with a matching key structure that requires validation.


We may consider a hard coded hbar value based on the piping options available to us.

Reply:
Ah, I wasn't clear.  I can charge a _hard coded_ amount, e.g., the 3000gas that `ECRECOVER` charges.  I can't charge some fee-schedule amount specific to this method.  I'll rewrite this.